### ISSUE TITLE
[Bug]: Service object type from the service object typedesc does not have the annotations

### ISSUE THREAD

**Issue by TharmiganK** [ballerina-platform/ballerina-lang#42906]
### Description

When we get the service object type from the service object typedesc, the annotations at the service level, resource level and resource parameter level are not visible at runtime

### Steps to Reproduce

I have the following service object type:
```bal
@http:ServiceContractConfig {
    basePath: "/social-media"
}
@http:ServiceConfig {
    auth: [
        {
            fileUserStoreConfig: {},
            scopes: ["admin"]
        }
    ]
}
public type Service service object {
    *http:ServiceContract;

    ...

    @http:ResourceConfig {
        name: "user",
        linkedTo: [
            {name: "user", method: http:DELETE, relation: "delete-user"},
            {name: "posts", method: http:POST, relation: "create-posts"},
            {name: "posts", method: http:GET, relation: "get-posts"}
        ]
    }
    resource function get users/[int id](@http:Header string X\-API\-Version) returns @http:Payload {mediaType: "application/org+json"} User|UserNotFound|error;
    
    ...
}
```

Here, I have annotations at service level, resource level and resource parameter level. I pass the typedesc of this service via an annotation - field name `serviceType`:
```bal
@http:ServiceConfig {
    serviceType:  Service
}
service Service "/social-media" on new http:Listener(8080) {
  
    ...

    resource function get users/[int id]() returns socialMedia:User|socialMedia:UserNotFound|error {
        return {id: 1, name: "John Doe", email: "johndoe@gmail.com"};
    }

    ...
}
```

At runtime, I can get the service object type as follows:
`BTypedesc` -> `ReferenceType` -> `ServiceType` (using TypeUtils APIs)

In that service type, none of the annotations are visible

### Affected Version(s)

Ballerina SwanLake Update 9

### Related area

-> Compilation

### Related issue(s) (optional)

This is a requirement for this proposal: https://github.com/ballerina-platform/ballerina-library/issues/6378

**Comment by chiranSachintha** [comment]
This issue occurs when a service calls the listener from the compiler before setting the annotations to the types. (We are not generating type descriptors for the service types. Therefore, we need to retain the annotations in the type for now). With the following example, we can generate the issue.
```ballerina
import ballerina/io;

type AnnotRecord record {|
    string value;
|};

type AnnotRecord1 record {|
    typedesc value;
|};

annotation AnnotRecord annot on type;
annotation AnnotRecord1 annot1 on service, type;



@annot {
    value: "T1"
}
public type Service service object {
    resource function get pet() returns string ;
};



listener Listener lst = new Listener(9090);

@annot1 {
    value: Service
}
service Service on lst  {
    resource function get pet() returns  string {
        return  "string";
    }
}

public class Listener {
    int port;
    public function 'start() returns error? {
    }

    public function gracefulStop() returns error? {
    }

    public function immediateStop() returns error? {
    }

    public function detach(service object {} s) returns error? {
    }

    public function attach(service object {} s, string[]? name = ()) returns error? {
        AnnotRecord1? annot = (typeof s).@annot1;
        io:println(annot);
        if annot is AnnotRecord1 {
            io:println((typeof Service).@annot1);
            io:println((typeof (annot.value)).@annot);
        }
    }

    public function init(int port) {
        self.port = port;
    }
}

```

**Comment by MaryamZi** [comment]
Is this a duplicate of https://github.com/ballerina-platform/ballerina-lang/issues/33594 then? Are they accessible in the start method?

We tried fixing the original issue back then, but deferred due to complexity of a workaround vs going with the typedesc change.

**Comment by MaryamZi** [comment]
Modified the code to access relevant annotations.

```ballerina
import ballerina/io;

type AnnotRecord record {|
    string value;
|};

type AnnotRecord1 record {|
    typedesc value;
|};

annotation AnnotRecord annot on type;
annotation AnnotRecord1 annot1 on service, type;

@annot {
    value: "T1"
}
public type Service service object {
    resource function get pet() returns string ;
};

listener Listener lst = new Listener(9090);

@annot1 {
    value: Service
}
service Service on lst  {
    resource function get pet() returns  string {
        return  "string";
    }
}

public class Listener {
    int port;
    Service? s = ();
    public function 'start() returns error? {
        Service s = <Service> self.s;
        AnnotRecord1? annot = (typeof s).@annot1;
        io:println("start: ", annot);
        if annot is AnnotRecord1 {
            io:println(Service.@annot);
            io:println(annot.value.@annot);
        }
    }

    public function gracefulStop() returns error? {
    }

    public function immediateStop() returns error? {
    }

    public function detach(Service s) returns error? {
    }

    public function attach(Service s, string[]? name = ()) returns error? {
        self.s = s;
        AnnotRecord1? annot = (typeof s).@annot1;
        io:println("attach: ", annot);
        if annot is AnnotRecord1 {
            io:println(Service.@annot);
            io:println(annot.value.@annot);
        }
    }

    public function init(int port) {
        self.port = port;
    }
}
```

```cmd
Running executable

attach: {"value":typedesc Service}


start: {"value":typedesc Service}
{"value":"T1"}
{"value":"T1"}
```

The annotation is available at `start`. Assuming this translates to native code also, is accessing annotations at `start` an option for now, @TharmiganK?

**Comment by TharmiganK** [comment]
> The annotation is available at `start`. Assuming this translates to native code also, is accessing annotations at `start` an option for now?

Will check on this

**Comment by chiranSachintha** [comment]
> Is this a duplicate of #33594 then? Are they accessible in the start method?
> 
> We tried fixing the original issue back then, but deferred due to complexity of a workaround vs going with the typedesc change.

Yes.

**Comment by TharmiganK** [comment]
@MaryamZi I moved the logic to `start` and the **service level annotations are now present** for the service types defined in the same package (if I defined the service type in a different package and import, the annotations are present in `attach` and `start`)

But the original issue is still there where I cannot get **the resource level annotations**. The following is a reduced ballerina code which represents my actual use case:
```bal
import ballerina/io;

type ResourceAnnotation record {
    string value;
};

type ParameterAnnotation record {
    string value;
};

type ServiceAnnotation record {
    string value?;
    typedesc<ServiceContract> serviceContractType?;
};

annotation ResourceAnnotation ResourceConfig on function;
annotation ParameterAnnotation ParamConfig on parameter;
annotation ServiceAnnotation ServiceConfig on service, type;

public type Service service object {
};

public type ServiceContract service object {
};

public type User record {|
    int id;
    string name;
    string email;
|};

@ServiceConfig {
    value: "users-service-contract"
}
public type UsersService service object {
    *ServiceContract;

    @ResourceConfig {value: "get-users"}
    resource function post users(@ParamConfig {value: "payload-param"} User user) returns User|error;
};

public class Listener {

    Service? s = ();

    public function 'start() returns error? {
        ServiceAnnotation? serviceConfig = (typeof self.s).@ServiceConfig;
        io:println("self-start: ", serviceConfig);
        if serviceConfig is ServiceAnnotation {
            io:println("self-start: ", UsersService.@ServiceConfig);
            typedesc<ServiceContract>? serviceContractType = serviceConfig.serviceContractType;
            io:println("self-start: ", serviceContractType);
            if serviceContractType is typedesc<ServiceContract> {
                io:println("self-start: ", serviceContractType.@ServiceConfig);
            }
        }
    }

    public function gracefulStop() returns error? {
    }

    public function immediateStop() returns error? {
    }

    public function detach(Service s) returns error? {
    }

    public function attach(Service s, string[]? name = ()) returns error? {
        self.s = s;
        ServiceAnnotation? serviceConfig = (typeof s).@ServiceConfig;
        io:println("self-attach: ", serviceConfig);
        if serviceConfig is ServiceAnnotation { 
            io:println("self-attach: ", UsersService.@ServiceConfig);
            typedesc<ServiceContract>? serviceContractType = serviceConfig.serviceContractType;
            io:println("self-attach: ", serviceContractType);
            if serviceContractType is typedesc<ServiceContract> {
                io:println("self-attach: ", serviceContractType.@ServiceConfig);
            }
        }
    }

    public function init(int port) {
    }
}

@ServiceConfig {
    serviceContractType: UsersService
}
service UsersService on new Listener(8080) {
    resource function post users(User user) returns User|error {
        return user;
    }
}
```

> **Note:** Please note the resource level and resource parameter level annotations in the service contract type - `UsersService`

**Comment by chiranSachintha** [comment]
This is a bug in the compiler. The issue is that we added those annotations to the annotation map, but at runtime, we didn't apply these annotations to the type associated with the resource functions and parameters. We only set the annotations which attached to the service type. (As a solution, we need to add those annotations to their respective types by calling the `processAnnotation` function.)

**Comment by MaryamZi** [comment]
Seems to be an issue at runtime rather. 

The check at https://github.com/ballerina-platform/ballerina-lang/blob/v2201.9.2/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/AnnotationUtils.java#L61 is no longer sufficient with type reference types.

**Comment by MaryamZi** [comment]
Let's use
- https://github.com/ballerina-platform/ballerina-lang/issues/33594 to track the original issue and
- https://github.com/ballerina-platform/ballerina-lang/issues/43122 to track the new issue (tagging this for U10)

**Comment by MaryamZi** [comment]
> Let's use
> 
> * [Annotation doesn't get populated before listener init #33594](https://github.com/ballerina-platform/ballerina-lang/issues/33594) to track the original issue and
> * [Method annotations not available on object type descriptors #43122](https://github.com/ballerina-platform/ballerina-lang/issues/43122) to track the new issue (tagging this for U10)

Closing this since this was/is tracked in other issues.
